flink-python/pyflink/table/types.py

```diff
@@ -15,172 +15,1803 @@
-    BYTE = ByteType()
-    CHAR = CharType()
-    SHORT = ShortType()
-    INT = IntegerType()
-    LONG = LongType()
-    FLOAT = FloatType()
-    DOUBLE = DoubleType()
-    DATE = DateType()
-    TIME = TimeType()
-    TIMESTAMP = TimestampType()
+    Creates a verifier that checks the type of obj against data_type and raises a TypeError if they
+    do not match.
+
+    This verifier also checks the value of obj against data_type and raises a ValueError if it's
+    not within the allowed range, e.g. using 128 as TinyIntType will overflow. Note that, Python
+    float is not checked, so it will become infinity when cast to Java float if it overflows.
+
+    >>> _create_type_verifier(RowType([]))(None)
+    >>> _create_type_verifier(VarCharType())("")
+    >>> _create_type_verifier(LongType())(0)
+    >>> _create_type_verifier(ArrayType(SmallIntType()))(list(range(3)))
+    >>> _create_type_verifier(ArrayType(VarCharType()))(set()) # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    TypeError:...
+    >>> _create_type_verifier(MapType(VarCharType(), IntType()))({})
+    >>> _create_type_verifier(RowType([]))(())
+    >>> _create_type_verifier(RowType([]))([])
+    >>> _create_type_verifier(RowType([]))([1]) # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    ValueError:...
+    >>> # Check if numeric values are within the allowed range.
+    >>> _create_type_verifier(TinyIntType())(12)
+    >>> _create_type_verifier(TinyIntType())(1234) # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    ValueError:...
+    >>> _create_type_verifier(TinyIntType(), False)(None) # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    ValueError:...
+    >>> _create_type_verifier(
+    ...     ArrayType(SmallIntType(), False))([1, None]) # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    ValueError:...
+    >>> _create_type_verifier(MapType(VarCharType(), IntType()))({None: 1})
+    Traceback (most recent call last):
+        ...
+    ValueError:...
+    >>> schema = RowType().add("a", IntType()).add("b", VarCharType(), False)
+    >>> _create_type_verifier(schema)((1, None)) # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    ValueError:...
+    """
+
+    if name is None:
+        new_msg = lambda msg: msg
+        new_name = lambda n: "field %s" % n
+    else:
+        new_msg = lambda msg: "%s: %s" % (name, msg)
+        new_name = lambda n: "field %s in %s" % (n, name)
+
+    def verify_nullability(obj):
+        if obj is None:
+            if data_type.nullable:
+                return True
+            else:
+                raise ValueError(new_msg("This field is not nullable, but got None"))
+        else:
+            return False
+
+    _type = type(data_type)
+
+    def assert_acceptable_types(obj):
```

Comment:
Maybe we can check the data type before verifier is created? The input `obj` is unnecessary for the core logic `assert _type in _acceptable_types`. We needn't run this function every time.

Reply:
Good catch. +1.